planner/optimize.go

```diff
@@ -411,8 +422,41 @@ func handleStmtHints(hints []*ast.TableOptimizerHint) (stmtHints stmtctx.StmtHin
 		case "nth_plan":
 			forceNthPlanCnt++
 			forceNthPlan = hint
+		case "set_var":
+			setVarHint := hint.HintData.(ast.HintSetVar)
+
+			// Not all session variables are permitted for use with SET_VAR
+			sysVar := variable.GetSysVar(setVarHint.VarName)
+			if sysVar == nil {
+				warn := terror.ClassUtil.New(errno.ErrUnresolvedHintName, fmt.Sprintf(errno.MySQLErrName[errno.ErrUnresolvedHintName], setVarHint.VarName, hint.HintName.O))
```

Comment:
how about defining this error to `planner/core/errors.go` like `ErrNotHintUpdatable`:

```go
errUnresolvedHintName = terror.ClassOptimizer.New(mysql.ErrUnresolvedHintName, mysql.MySQLErrName[mysql.ErrUnresolvedHintName]
errNotHintUpdatable = terror.ClassOptimizer.New(mysql.ErrNotHintUpdatable, mysql.MySQLErrName[mysql.ErrNotHintUpdatable])
```

Then we can use this error to generate a new error with a call stack:

```go
warns = append(warns, errUnresolvedHintName.GenWithStackByArgs(setVarHint.VarName, hint.HintName.O))
warns = append(warns, errNotHintUpdatable.GenWithStackByArgs(setVarHint.VarName))
```